plugin/storage/cassandra/spanstore/reader.go

```diff
@@ -177,26 +240,119 @@ func (s *SpanReader) FindTraces(queryParams *spanstore.TraceQueryParameters) ([]
 	return retMe, nil
 }
 
-// executeQuery returns UniqueTraceIDs
-func (s *SpanReader) executeQuery(query Query, tableMetrics *casMetrics.Table) (dbmodel.UniqueTraceIDs, error) {
+func (s *SpanReader) findTraceIDs(traceQuery *spanstore.TraceQueryParameters) (dbmodel.UniqueTraceIDs, error) {
+	if traceQuery.DurationMin != 0 {
+		return s.queryByDuration(traceQuery)
+	}
+
+	// NB(rooz): consider refactoring the query... methods where they take in traceQuery instead of a breakdown of its values
+	if traceQuery.OperationName != "" {
+		traceIds, err := s.queryByServiceNameAndOperation(traceQuery)
+		if err != nil {
+			return nil, err
+		}
+		if len(traceQuery.Tags) > 0 {
+			tagTraceIds, err := s.queryByTagsAndLogs(traceQuery)
+			if err != nil {
+				return nil, err
+			}
+			return dbmodel.IntersectTraceIDs([]dbmodel.UniqueTraceIDs{
+				traceIds,
+				tagTraceIds,
+			}), nil
+		}
+		return traceIds, err
+	}
+	if len(traceQuery.Tags) > 0 {
+		return s.queryByTagsAndLogs(traceQuery)
+	}
+	return s.queryByService(traceQuery)
+}
+
+func (s *SpanReader) queryByTagsAndLogs(tq *spanstore.TraceQueryParameters) (dbmodel.UniqueTraceIDs, error) {
+	results := dbmodel.UniqueTraceIDs{}
+	for k, v := range tq.Tags {
+		query := s.session.Query(
+			queryByTag,
+			tq.ServiceName,
+			k,
+			v,
+			model.TimeAsEpochMicroseconds(tq.StartTimeMin),
+			model.TimeAsEpochMicroseconds(tq.StartTimeMax),
+			tq.NumTraces,
+		)
+		t, err := s.executeQuery(query, s.metrics.queryTagIndex)
+		if err != nil {
+			return nil, err
+		}
+		// (NB)rooz: I don't think this is the behavior we actually want. This is OR over queried tags, do we want that or AND behavior?
```

Comment:
I think we want AND. You can switch to AND by using Intersect() function instead of adding to the result set.